compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -938,6 +940,35 @@ impl<'a> Parser<'a> {
             && self.look_ahead(1, |tok| tok.kind == TokenKind::OpenDelim(Delimiter::Parenthesis))
             && let Some(path) = self.recover_path_from_fn()
         {
+            path
+        } else if !self.token.is_path_start() && self.token.can_begin_type() && let Ok(ty) = self.parse_ty_no_plus() {
```

Comment:
The issue I'm seeing is that `let Ok(ty) = self.parse_ty_no_plus()` can fail parsing the type and will be returning an `Err(Diagnostic)`, which gets discarded. If this happens, an ICE will occur. We need to either bubble it up (wich `self.parse_ty_no_plus()?`, which is fine if slightly misleading) or `err.delay_as_bug()` it (which isn't great without another mechanism to complain about the parse failure).

Reply:
Hmm, what about matching on error and cancelling the diagnostic. I didn't properly realize that it returned a diagnostic, not just an error to be later turned into a diagnostic.

Reply:
With my current code the following ICEs:
```rust
fn foo<T: &match>() {}
```
if we `?` the parse call then we get:
```text
error: expected type, found keyword `match`
 --> <anon>:1:12
  |
1 | fn foo<T: &match>() {}
  |            ^^^^^ expected type
```
I'm happy with this, I don't think it's confusing. We could however add a note, something along the lines of: `this error was encountered when trying to recover from a "expected a trait, found type" error` before bubbling it up.

Reply:
You can do `map_err(|mut err| { err.note(..); err })?`, but I think even without the extra context this output would be fine.